### PR TITLE
fix: Handle trailing slash in routes when building StaticRoute's

### PR DESCRIPTION
**Pull request by Tehnix**
In https://github.com/leptos-rs/leptos/issues/2663 I ran into an issue with hydration for sub-routes. While I was initially a bit off, after @gbj's comment I was able to fix the route I was generating but then ran into another issue with `build_static_routes` not supporting paths ending in a trailing slash (e.g. `another-page/`).

It would instead think these paths had no trailing slash, and when combined with the `TrailingSlash::Exact` setting, you would see output like this

```
building static route: /another-page
Errors: [
    NotFound,
]
building static route:
```

This PR does a couple of things:

- Changes `spawn_task` to return a handle that you can await on
  - This allows `build_static_routes` to return when its done generating the files
  - It then enables the user to do any file manipulation they need in their `main` function after calling `build_static_routes` (previously any attempt to do this would break since the files were generated async and the `.await` on `build_static_routes` wouldn't actually wait for the files to be generated)
- Adds `trailing_slash` to `StaticPath` to track `/` route endings
- Retains the trailing slash when turning the `StaticPath` into a `ResolvedStaticPath`, making sure the correct route gets built
- Changes the file output to be `<route>/index.html` for routes ending with a trailing slash

I've also extended the output a bit, which makes it easier for users to follow along in what route corresponds to what file, written after the current `building static route`:

```
...
building static route: /sub-page/
writing static route to: target/site/sub-page/index.html
building static route: /another-page/
writing static route to: target/site/another-page/index.html
building static route:
writing static route to: target/site/index.html
```

I'm rather new to Rust and especially new to the Leptos codebase, so if there's any improvements I can make I'd love to hear!

I've tested it locally, and it completely fixes https://github.com/leptos-rs/leptos/issues/2663 :)

### CODE REVIEW

File: integrations/axum/src/lib.rs [hunk 1]
```diff
@@ -1275,6 +1278,7 @@ pub async fn build_static_routes<IV>(
         .await
         .expect("could not build static routes")
     });
+    let _ = handle.await;
```
Comment (gbj) [inline integrations/axum/src/lib.rs:1281]: `leptos_reactive::spawn_local` (like the underlying [`wasm_bindgen_futures::spawn_local`](https://docs.rs/wasm-bindgen-futures/latest/wasm_bindgen_futures/fn.spawn_local.html)) returns `()`, not a handle that can be awaited, so you can see the following error in CI when the `wasm` feature is enabled:
```
 error[E0277]: `()` is not a future
    --> integrations/axum/src/lib.rs:1281:20
     |
1281 |     let _ = handle.await;
     |                   -^^^^^
     |                   ||
     |                   |`()` is not a future
     |                   help: remove the `.await`
     |
     = help: the trait `futures::Future` is not implemented for `()`, which is required by `(): std::future::IntoFuture`
     = note: () must be a future or must implement `IntoFuture` to be awaited
     = note: required for `()` to implement `std::future::IntoFuture`
```

**Review by gbj — CHANGES_REQUESTED** [review]
This mostly looks good to me but the changes to spawning cause it not to compile with the `wasm` feature.

File: integrations/axum/src/lib.rs (hunk 1 above)
Comment (Tehnix) [reply integrations/axum/src/lib.rs:1283]: Hadn't seen that CI was finally running 🥳 

Hmm, while I've got rid of all the errors, I did end up with a bit of an iffy solution to it, essentially:

```
    #[allow(unused_variables, clippy::let_unit_value)]
    let handle = spawn_task!(async move {
        leptos_router::build_static_routes(
            &options,
            app_fn,
            &routes,
            &static_data_map,
        )
        .await
        .expect("could not build static routes")
    });
    #[cfg(all(not(feature = "wasm"), feature = "default"))]
    let _ = handle.await;
```

I'd be curious if you have any suggestions to alternatives? For now I've been trying to be minimally invasive, but that might be the wrong approach here 🤔

**Comment by gbj** [comment]
I have since merged 0.7 into the main branch. 0.7 does not yet have static routing implemented. I will make sure the case this addresses is working once that is implemented.

**Comment by Tehnix** [comment]
> I have since merged 0.7 into the main branch. 0.7 does not yet have static routing implemented. I will make sure the case this addresses is working once that is implemented.

Awesome, very excited for 0.7! 🥳 Let me know if you end up with wanting help on it (at the very least I can help test).

@gbj should I close this PR? There's still https://github.com/leptos-rs/leptos/issues/2663 to track the actual issue itself. (feel free to close it if you feel that's appropriate)

**Comment by gbj** [comment]
This is handled by #2875 so should work as intended in the new version.

**Pull request closed without merging**